src/crypto/sodium.cpp

```diff
@@ -22,138 +22,139 @@
+                unsigned char* sig)
+{
+    if (!sig || !msg)
+    {
+        return 0;
     }
-    free(sigbuffer);
-    free(pubKey);
-    free(privKey);
-    return(bufferlen);
+
+    int result = crypto_sign_detached(sig, NULL, msg, msglen, (const unsigned char*)privKey);
+
+    return (result == 0) ? (crypto_sign_BYTES + msglen) : 0;
 }
 
 
 // Verifies the signature of a message.
 int EdDSA::verify(const unsigned char* msg, unsigned long long msglen,
-                  const unsigned char* sig, const unsigned char* pubKey) {
-    unsigned char* msgbuffer = (unsigned char*)malloc(msglen + crypto_sign_BYTES);
-    if (msgbuffer == NULL) {
-        // Something went wrong allocating the memory.
-        return(0);
-    }
-    unsigned char* signedmsg = (unsigned char*)malloc(msglen + crypto_sign_BYTES);
-    if (signedmsg == NULL) {
-        // Something went wrong allocating the memory.
-        free(msgbuffer);
-        return(0);
+                  const unsigned char* sig, const unsigned char* pubKey)
+{
+    if (!sig || !msg)
+    {
+        return 0;
     }
 
-    // Assemble signed message in one array.
-    memcpy(signedmsg, msg, msglen);
-    memcpy(signedmsg + msglen, sig, crypto_sign_BYTES);
-
-    unsigned long long bufferlen = 0;
-    int result = crypto_sign_open(msgbuffer, &bufferlen,
-                                  signedmsg, msglen + crypto_sign_BYTES, pubKey);
-    free(msgbuffer);
-    free(signedmsg);
-    if (result) {
-        return(1);
-    } else {
-        // crypto_sign_open() returns -1 on failure.
-        return(0);
-    }
+    int result = crypto_sign_verify_detached(sig, msg, msglen, pubKey);
+
+    return (result == 0) ? 1 : 0;
 }
 
+const byte *EdDSA::genFingerprint()
+{
+    HashSHA256 hash;
+    string binaryhash;
+    hash.add((byte*)&pubKey, PUBLIC_KEY_LENGTH);
+    hash.get(&binaryhash);
 
-// Generates a new Ed25519 private key seed. The key seed is stored in the object.
-int EdDSA::genKeySeed(unsigned char* privKey) {
-    // Make space for a new key seed (if not present).
-    if (!this->keySeed) {
-        this->keySeed = (unsigned char*)malloc(crypto_sign_SEEDBYTES);
-        if (this->keySeed == NULL) {
-            // Something went wrong allocating the memory.
-            return(0);
-        }
-    }
-    // Now make the new key seed.
-    this->rng.GenerateBlock(this->keySeed, crypto_sign_SEEDBYTES);
-    // Copy it to privKey before returning.
-    if (privKey)
+    return (byte*) binaryhash.substr(0, 20).c_str();
+}
+
+const char *EdDSA::genFingerprintHex()
+{
+    const byte *fp = genFingerprint();
+
+    static const char hexchars[] = "0123456789abcdef";
+    ostringstream oss;
+    string fpHex;
+    for (size_t i = 0; i < 20; ++i)
     {
-        memcpy(privKey, this->keySeed, crypto_sign_SEEDBYTES);
+        oss.put(hexchars[(fp[i] >> 4) & 0x0F]);
+        oss.put(hexchars[fp[i] & 0x0F]);
     }
-    return(1);
+    fpHex = oss.str();
+
+    return fpHex.c_str();
```

Comment:
It's returning a pointer that belongs to a local variable.
It will be invalid after the call to the function.